articles/understanding-node-js-async-control-flows/index.md

```diff
@@ -0,0 +1,144 @@
+
+**Note:** if your task has no timers or does not perform any I/O, the code will run in series, and every task will happen one after the other. If one of the functions returns an error to its callback. The final callback will immediately be invoked to null or with the value of the error output.
+
+[Race](https://caolan.github.io/async/v3/docs.html#race) is a close relative to parallel. `async.race` runs a collection of tasks parallel. There is a small difference between race and parallel. As soon as any given function executes successfully, the main callback is invoked immediately with the results of that function.
+
+##### Parameters
+
+- Tasks - an array containing multiple tasks to run asynchronously, every function can execute successfully with an optimal result value.
+- Callback - function that runs once the tasks are successfully executed or when one on the tasks returns an error. In a race, this function will get the results or an error of the first function that executes successfully.
+
+**Syntax:** `async.race(tasks, callback)`
+
+<iframe height="400px" width="100%" src="https://repl.it/@kimkimani/async-race?lite=true" scrolling="no" frameborder="no" allowtransparency="true" allowfullscreen="true" sandbox="allow-forms allow-pointer-lock allow-popups allow-same-origin allow-scripts allow-modals"></iframe>
+
+#### Queue
+
+A [queue](https://caolan.github.io/async/v3/docs.html#queue) can be used to run tasks asynchronously. A queue is created with a specified concurrency. Queue tasks are executed parallel but with a concurrency limit. A queue object is completed based on an asynchronous operation and passed to a worker. If all workers are in progress, the task is queued until one becomes available. When a worker completes a task, the task callback is called.
+
+#### Parameters
+
+- Worker - workers are invoked with tasks and callbacks. Worker is an asynchronous function. It processes all tasks assigned to the queue. Tasks and callbacks are the main parameters processed by a worker.
+- Concurrency - determines the number of worker functions to run parallel during the lifecycle of a queue.
+
+**Syntax:** `async.queue (tasks, concurrency)`
+
+`async.queue` returns a queue object that manages a task. These object properties can be attached to callbacks to listen to specific tasks during the queue lifecycle.
+These [queue objects](https://caolan.github.io/async/v3/docs.html#QueueObject) include.
+
+- Push - it is an async function that adds a task to a queue to be processed within the workers. Once the workers finish processing these tasks, they are called within a callback, which can take a single task or an array of tasks. A push is invoked with a `queue.push(task, [callback])`.
+- Drain - it specifies a callback to be called after the last task item of the queue has finished and passed to a worker. Invoked with a `q.drain()` 
```

Comment:
I had a question here, is it `q.drain()` or `queue.drain`

Reply:
`q.drain()` yeah. The code example has it used